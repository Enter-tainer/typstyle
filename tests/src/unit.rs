use std::{env, error::Error, fs, path::Path};

use insta::internals::Content;
use libtest_mimic::{Failed, Trial};
use typst_syntax::Source;
use typstyle_core::Typstyle;

use crate::common::{fixtures_dir, read_source_with_options};

/// Creates one test for each `.typ` file in the current directory or
/// sub-directories of the current directory.
pub fn collect_tests() -> Result<Vec<Trial>, Box<dyn Error>> {
    fn make_snapshot_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_snapshot(&path, width)
        })
        .with_kind("snapshot")
    }

    fn make_convergence_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_convergence(&path, width)
        })
        .with_kind("convergence")
    }

    #[cfg(feature = "consistency")]
    fn make_consistency_test(path: &Path, name: &str, width: usize) -> Trial {
        let path = path.to_path_buf();
        Trial::test(format!("{name} - {width}char"), move || {
            check_output_consistency(&path, width)
        })
        .with_kind("consistency")
    }

    fn visit_dir(path: &Path, tests: &mut Vec<Trial>) -> Result<(), Box<dyn Error>> {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let file_type = entry.file_type()?;
            let path = entry.path();

            if file_type.is_dir() {
                // Handle directories
                if path.file_name() == Some("partial".as_ref()) {
                    // Ignore partial tests
                    continue;
                }
                visit_dir(&path, tests)?;
                continue;
            } else if !(file_type.is_file() && path.extension() == Some("typ".as_ref())) {
                continue;
            }
            // Handle .typ files
            let name = path
                .strip_prefix(env::current_dir()?)?
                .display()
                .to_string();

            let is_no_snap = path
                .file_stem()
                .and_then(|name| name.to_str())
                .is_some_and(|name| name.starts_with('_'));
            if !is_no_snap {
                tests.extend([
                    make_snapshot_test(&path, &name, 0),
                    make_snapshot_test(&path, &name, 40),
                    make_snapshot_test(&path, &name, 80),
                    make_snapshot_test(&path, &name, 120),
                ]);
            }
            tests.extend([
                make_convergence_test(&path, &name, 0),
                make_convergence_test(&path, &name, 40),
                make_convergence_test(&path, &name, 80),
            ]);
            #[cfg(feature = "consistency")]
            tests.extend([
                make_consistency_test(&path, &name, 0),
                make_consistency_test(&path, &name, 40),
                make_consistency_test(&path, &name, 80),
            ]);
        }

        Ok(())
    }

    // We recursively look for `.typ` files, starting from the current directory.
    let mut tests = Vec::new();
    let current_dir = fixtures_dir();
    visit_dir(&current_dir, &mut tests)?;

    Ok(tests)
}

fn check_snapshot(path: &Path, width: usize) -> Result<(), Failed> {
    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;

    let mut settings = insta::Settings::clone_current();
    settings.set_prepend_module_to_snapshot(false);
    settings.set_omit_expression(true);
    settings.set_snapshot_path(path.parent().unwrap().join("snap"));
    settings.set_input_file(path);
    if source.root().erroneous() {
        settings.set_raw_info(&Content::Map(vec![("erroneous".into(), true.into())]));
    }
    settings.bind(|| {
        let snap_name = format!("{}-{width}", path.file_name().unwrap().to_str().unwrap());
        if source.root().erroneous() {
            insta::assert_snapshot!(snap_name, "");
        } else {
            cfg.max_width = width;
            let mut formatted = Typstyle::new(cfg).format_source(&source).unwrap();
            if formatted.starts_with('\n') {
                formatted.insert_str(0, "// DUMMY\n");
            }
            if formatted.ends_with("\n\n") {
                formatted.push_str("// DUMMY\n");
            }

            insta::assert_snapshot!(snap_name, formatted);
        }
    });
    Ok(())
}

fn check_convergence(path: &Path, width: usize) -> Result<(), Failed> {
    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;
    if source.root().erroneous() {
        return Ok(());
    }

    cfg.max_width = width;
    let mut first_pass = Typstyle::new(cfg.clone()).format_source(&source)?;
    for i in 0..=opt.relax_convergence {
        let new_source = Source::detached(&first_pass);
        if new_source.root().erroneous() {
            panic!(
                "the source becomes erroneous after {} iterations:\n{:#?}",
                i + 1,
                new_source.root().errors()
            )
        }
        let second_pass = Typstyle::new(cfg.clone()).format_source(&new_source)?;
        if first_pass == second_pass {
            return Ok(());
        }
        if i == opt.relax_convergence {
            pretty_assertions::assert_str_eq!(
                first_pass,
                second_pass,
                "formatting does not converge in {} iterations!",
                opt.relax_convergence
            );
        }
        first_pass = second_pass;
    }

    Ok(())
}

#[cfg(feature = "consistency")]
fn check_output_consistency(path: &Path, width: usize) -> Result<(), Failed> {
    use std::path::PathBuf;

    use typstyle_consistency::{ErrorSink, FormattedSources, FormatterHarness};

    let (source, opt) = read_source_with_options(path)?;
    let mut cfg = opt.config;
    if source.root().erroneous() {
        return Ok(());
    }

    cfg.max_width = width;

    let mut err_sink = ErrorSink::new(format!("consistency {}", path.display()));

    let mut harness = FormatterHarness::new("".to_string(), PathBuf::new())?;
    let main_vpath = Path::new("__main__");
    harness.add_source_file(main_vpath, source.text())?;

    let base_world = harness.snapshot();
    let fmt_sources = FormattedSources {
        name: "formatted".to_string(),
        sources: harness.format(
            &base_world,
            |source| Ok(Typstyle::new(cfg.clone()).format_source(&source).unwrap()),
            &mut err_sink,
        )?,
    };

    harness.compile_and_compare([fmt_sources].iter(), main_vpath, false, &mut err_sink)?;

    if err_sink.is_ok() {
        Ok(())
    } else {
        Err(err_sink.into())
    }
}
