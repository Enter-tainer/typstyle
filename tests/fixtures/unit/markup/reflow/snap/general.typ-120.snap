---
source: tests/src/unit.rs
input_file: tests/fixtures/unit/markup/reflow/general.typ
---
/// typstyle: wrap_text

Lorem ipsum dolor sit amet, consectetur adipiscing elit. In vel justo a massa consequat fermentum et nec tellus.

Vestibulum pulvinar neque a pharetra tincidunt. Aenean commodo neque non ipsum accumsan, vitae vehicula sapien mattis.


Proin ornare arcu a ligula vulputate, in interdum leo tincidunt. Sed nec massa rhoncus nibh luctus congue. Ut in velit
non enim viverra feugiat in et risus. Morbi pharetra eros ac justo porta imperdiet. Mauris venenatis ipsum et luctus
vehicula.

---

Here is some normal text followed by an equation: $x^2 + y^2 = z^2$ and more text.

This line contains *strong text* mixed with `raw code` and regular words.

Let's add a complex equation:
$ integral_0^infinity e^(-x^2) dif x = sqrt(pi)/2 $
surrounded by text.

Testing _multiple_ `inline` $"elements"$ in a single line with various _spacing_ patterns.

Some text with equations $E = m c^2$ and $F = m a$ inline with _emphasized words_ and `code snippets`.

Here's a paragraph with line breaks and multiple spaces to test reflow behavior with $"complex" = (a + b)^2$ equations.

Technical terms like ```ts Array<T>``` mixed with *important notes* and $sum_(i=1)^n x_i$.

// Testing comments and special characters Let's include #hashtags and @mentions with $theta$ and `const`.

- Simple list item with *bold text* and $alpha$ equation
- Multi-line list item with `code` and continuation with extra spaces
- List with equation: $f(x) = sin(x)$ and _emphasis_

/ Term: Definition with *strong* and $beta$ mixed
/ Complex Term:
  Multi-line definition with `code block` and $integral$ equation
/ Technical: Uses ```ts Array<T>``` with *important* notes about $theta$

+ First enumerated item with `inline code` and $x^2$ equation
+ Second item with *bold* and multiple spaces
+ Mixed content: _emphasis_, $pi$, and `const`

- Nested items Just nested items:
  - Sub-item with $alpha$ and *bold*
  - Another with `code` and line breaks
  - Mixed _formatting_ and $beta$ here

- List item with #strong[bold text inside function] and regular text
- Item with nested functions #emph[emphasized with #strike[strikethrough]] and continuation

/ Term with functions: Using #text(blue)[colored text] with $alpha$ formula
/ Complex function:
  #table(
    columns: (auto, auto),
    [*Header 1*], [*Header 2*],
    [Content], [More #emph[styled] content],
  )
  Text after function call.

+ Enumerated with #link("https://example.com")[link text] and `code`
+ Item using #text(size: 1.2em)[larger text] with spacing
+ Mixed #strong[formatted] and #emph[emphasized] content with $E = m c^2$

- #block[
    Multi-line block content with nested #strong[formatting] and extra spaces
  ]

/ Functions with parameters: Definition using #text(weight: "bold", size: 12pt)[Formatted text with spaces] and
  $"equations"$

= Heading with *mixed* `formatting` and $"math"$ elements
Text between headings.
== Sub-heading with #text(blue)[colored] and #text(style: "italic")[styled] text

=== Another heading
Multi-paragraph list item:
- This is a list item with multiple paragraphs.

  Second paragraph in the same list item with irregular spacing.

  Third paragraph with *strong*, _emphasized_, and `code` elements.

- List item with a footnote#footnote[This is a footnote with *formatting* and $"equations"$ that should correctly
    reflow] and continuation.

/ Term with long content:
  This is a very long description that will likely need to wrap when formatted properly. It contains *strong text*,
  `inline code`, and $"math" = "elements"$ with various spacing patterns to test how reflow handles complex term
  definitions that extend beyond the normal line length.

#figure(
  ```python
  def example():
      # This is a code block with syntax highlighting
      return "test" + " " + "string"
  ```,
  caption: [Figure with *formatted* caption and $"math"$ elements],
)

Multi-paragraph markup with complex elements:
- List item with #text(fill: rgb(10%, 20%, 30%))[
    Multi-line content inside a function with nested #strong[formatting] and equations $sum_(i=1)^n x_i^2$
  ]

- Item with #grid(
    columns: (1fr, 1fr),
    [Cell with *formatting*], [Another cell with $"math"$ and `code`],
  )

/ Term with display math:
  $ sum_(i=0)^n i = frac(n(n+1), 2) $
  And explanation with more text following
Text with special Unicode: "Em—dash" and "en–dash" with *formatting* mixed in áéíóú äëïöü characters.

This paragraph contains special sequences that shouldn't be interpreted as markup: A mathematical expression like 3 +
4 = 7 should remain intact. Symbol sequences like a / b or x - y should not be treated as list markers.

URLs in definitions:
/ Docs: Find them at https://docs.typst.app/
/ Resources: https://typst.app/ and forums at https://community.typst.app/

This might be yet a bit cleaner.
#highlight[Actually, it's muuuuuch worse in performance. Moreover the first version can be optimized to compute many
  marks at once.
]
